crates/lib/src/execution_data.rs

```diff
@@ -1,60 +1,114 @@
 
 impl ReadoutMap {
-    /// Given a known readout field name and index, return the result's [`ReadoutValues`], if any.
+    /// Returns a [`ReadoutValue`] for the given memory index and shot number, if any
     #[must_use]
-    pub fn get_readout_values(&self, field: String, index: u64) -> Option<ReadoutValues> {
-        let readout_values = self.0.get(&MemoryReference { name: field, index })?;
+    pub fn get_value(
+        &self,
+        register_name: &str,
+        index: usize,
+        shot_num: usize,
+    ) -> Option<ReadoutValue> {
+        self.0
+            .get(register_name)
+            .and_then(|matrix| matrix.get((index, shot_num)))
+            .copied()?
+    }
 
-        Some(readout_values.clone())
+    /// Returns a vector of the [`ReadoutValue`]s in the given register at a particular memory
+    /// index across all shots.
+    #[must_use]
+    pub fn get_values_by_memory_index(
+        &self,
+        register_name: &str,
+        index: usize,
+    ) -> Option<Vec<Option<ReadoutValue>>> {
+        let register = self.0.get(register_name);
+        if let Some(matrix) = register {
+            if index >= matrix.nrows() {
+                return None;
+            }
+            Some(matrix.row(index).to_vec())
+        } else {
+            None
+        }
     }
 
-    /// Given a known readout field name, return the result's [`ReadoutValues`] for all indices, if any.
-    pub fn get_readout_values_for_field(
+    /// Returns a vector of the [`ReadoutValue`]s in the given register for a particular shot number.
+    #[must_use]
+    pub fn get_values_by_shot(
         &self,
-        field: &str,
-    ) -> Result<Option<Vec<Option<ReadoutValues>>>, TryFromIntError> {
-        let mut readout_values = Vec::new();
-        for (memref, values) in &self.0 {
-            let MemoryReference { name, index } = memref;
-            let index = usize::try_from(*index)?;
-            if name == field {
-                if readout_values.len() <= index {
-                    readout_values.resize(index + 1, None);
-                }
-                readout_values[index] = Some(values.clone());
+        register_name: &str,
+        shot: usize,
+    ) -> Option<Vec<Option<ReadoutValue>>> {
+        let register = self.0.get(register_name);
+        if let Some(matrix) = register {
+            if shot >= matrix.ncols() {
+                return None;
             }
+            Some(matrix.column(shot).to_vec())
+        } else {
+            None
         }
+    }
 
-        Ok((!readout_values.is_empty()).then_some(readout_values))
+    /// Returns the matrix as a 2-dimensional array where the row is the memory index and the
+    /// column is the shot number.
+    #[must_use]
+    pub fn get_index_wise_matrix(&self, register_name: &str) -> Option<&RegisterMatrix> {
+        self.0.get(register_name)
+    }
+
+    /// Returns the matrix as a 2-dimensional array where the row is the shot number, and the
```

Comment:
What matrix?

Reply:
Updated the docstring, and made it reference `RegisterMatrix`